File: src/Microsoft.VisualBasic.Core/tests/VB/SpecialDirectoriesTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.
Option Explicit On
Option Strict On

Imports System
Imports System.Environment
Imports Microsoft.VisualBasic.FileIO
Imports Xunit
Namespace Microsoft.VisualBasic.Tests.VB
    Public NotInheritable Class SpecialDirectoriesTests

        Private Shared Sub CheckSpecialFolder(folder As SpecialFolder, getSpecialDirectory As Func(Of String))
            Dim LocalFolderPath As String = Environment.GetFolderPath(folder)
            If LocalFolderPath = "" Then
                Assert.Throws(Of IO.DirectoryNotFoundException)(Function() getSpecialDirectory.Invoke)
            Else
                Assert.Equal(LocalFolderPath.TrimSeparators, getSpecialDirectory().TrimSeparators)
            End If

        End Sub

        <Fact>
        Public Shared Sub AllUsersApplicationDataFolderTest()
            Assert.Throws(Of PlatformNotSupportedException)(Function() SpecialDirectories.AllUsersApplicationData)
        End Sub

        <Fact>
        Public Shared Sub CurrentUserApplicationDataFolderTest()
            Assert.Throws(Of PlatformNotSupportedException)(Function() SpecialDirectories.CurrentUserApplicationData)
        End Sub

        <Fact>
        Public Shared Sub DesktopFolderTest()
            CheckSpecialFolder(SpecialFolder.Desktop, Function() SpecialDirectories.Desktop)
        End Sub

        <Fact>
        Public Shared Sub MyDocumentsFolderTest()
            If PlatformDetection.IsWindowsNanoServer Then
                Assert.Throws(Of IO.DirectoryNotFoundException)(Function() SpecialDirectories.MyDocuments)
                Exit Sub
            End If

            CheckSpecialFolder(SpecialFolder.MyDocuments, Function() SpecialDirectories.MyDocuments)
        End Sub

        <Fact>
        Public Shared Sub MyMusicFolderTest()
            CheckSpecialFolder(SpecialFolder.MyMusic, Function() SpecialDirectories.MyMusic)
        End Sub

        <Fact>
        Public Shared Sub MyPicturesFolderTest()
            CheckSpecialFolder(SpecialFolder.MyPictures, Function() SpecialDirectories.MyPictures)
        End Sub

        <Fact>
        Public Shared Sub ProgramFilesFolderTest()
            CheckSpecialFolder(SpecialFolder.ProgramFiles, Function() SpecialDirectories.ProgramFiles)
        End Sub

        <Fact>
        Public Shared Sub ProgramsFolderTest()
            CheckSpecialFolder(SpecialFolder.Programs, Function() SpecialDirectories.Programs)
        End Sub

        <Fact>
        Public Shared Sub TempFolderTest()
            Assert.Equal(IO.Path.GetTempPath.TrimSeparators, SpecialDirectories.Temp.TrimSeparators)
        End Sub

    End Class
End Namespace

File: src/Microsoft.VisualBasic.Core/tests/Microsoft/VisualBasic/FileIO/SpecialDirectoriesTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using static System.Environment;
using Xunit;

namespace Microsoft.VisualBasic.FileIO.Tests
{
    public class SpecialDirectoriesTests
    {
        private static void CheckSpecialFolder(SpecialFolder folder, Func<string> getSpecialDirectory)
        {
            var path = Environment.GetFolderPath(folder);
            if (string.IsNullOrEmpty(path))
            {
                Assert.Throws<System.IO.DirectoryNotFoundException>(getSpecialDirectory);
            }
            else
            {
                Assert.Equal(TrimSeparators(path), TrimSeparators(getSpecialDirectory()));
            }
        }

        [Fact]
        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Supported on netfx")]
        public static void AllUsersApplicationDataFolderTest()
        {
            Assert.Throws<System.IO.DirectoryNotFoundException>(() => SpecialDirectories.AllUsersApplicationData);
        }

        [Fact]
        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Supported on netfx")]
        public static void CurrentUserApplicationDataFolderTest()
        {
            Assert.Throws<System.IO.DirectoryNotFoundException>(() => SpecialDirectories.CurrentUserApplicationData);
        }

        [Fact]
        public static void DesktopFolderTest()
        {
            CheckSpecialFolder(SpecialFolder.Desktop, () => SpecialDirectories.Desktop);
        }

        [Fact]
        public static void MyDocumentsFolderTest()
        {
            if (PlatformDetection.IsWindowsNanoServer)
            {
                Assert.Throws<System.IO.DirectoryNotFoundException>(() => SpecialDirectories.MyDocuments);
            }
            else
            {
                CheckSpecialFolder(SpecialFolder.MyDocuments, () => SpecialDirectories.MyDocuments);
            }
        }

        [Fact]
        public static void MyMusicFolderTest()
        {
            CheckSpecialFolder(SpecialFolder.MyMusic, () => SpecialDirectories.MyMusic);
        }

        [Fact]
        public static void MyPicturesFolderTest()
        {
            CheckSpecialFolder(SpecialFolder.MyPictures, () => SpecialDirectories.MyPictures);
        }

        [Fact]
        public static void ProgramFilesFolderTest()
        {
            CheckSpecialFolder(SpecialFolder.ProgramFiles, () => SpecialDirectories.ProgramFiles);
        }

        [Fact]
        public static void ProgramsFolderTest()
        {
            CheckSpecialFolder(SpecialFolder.Programs, () => SpecialDirectories.Programs);
        }

        [Fact]
        public static void TempFolderTest()
        {
            // On Nano Server >=1809 the temp path's case is changed during the normalization.
            Assert.Equal(TrimSeparators(System.IO.Path.GetTempPath()), TrimSeparators(SpecialDirectories.Temp), ignoreCase: PlatformDetection.IsWindowsNanoServer);
        }

        private static string TrimSeparators(string s)
        {
            return s.TrimEnd(System.IO.Path.DirectorySeparatorChar, System.IO.Path.AltDirectorySeparatorChar);
        }
    }
}
